com.calibre_ebook.calibre.yaml

```diff
@@ -21,19 +21,30 @@ finish-args:
 modules:
   - name: calibre
     buildsystem: simple
+    build-options:
+      env:
+        - XDG_UTILS_INSTALL_MODE=system
     build-commands:
-      - mkdir -p /app/bin
-      - tar -x -f calibre*.txz -C /app/bin
+      # Extract Calibre archive to target location
+      - mkdir -p "${FLATPAK_DEST}/lib/calibre"
+      - tar -xvf calibre*.txz -C "${FLATPAK_DEST}/lib/calibre"
       - rm -f calibre*.txz
-      - install -Dm644 /app/bin/resources/images/lt.png /app/share/icons/hicolor/256x256/apps/com.calibre_ebook.calibre.png
-      - install -Dm644 /app/bin/resources/images/tweak.png /app/share/icons/hicolor/256x256/apps/com.calibre_ebook.calibre.ebook_edit.png
-      - install -Dm644 /app/bin/resources/images/viewer.png /app/share/icons/hicolor/256x256/apps/com.calibre_ebook.calibre.ebook_viewer.png
-      - install -Dm644 /app/bin/resources/images/viewer.png /app/share/icons/hicolor/256x256/apps/com.calibre_ebook.calibre.lrfviewer.png
-      - install -Dm644 -t /app/share/metainfo com.calibre_ebook.calibre.appdata.xml
-      - install -Dm644 -t /app/share/applications com.calibre_ebook.calibre.desktop
-      - install -Dm644 -t /app/share/applications com.calibre_ebook.calibre.ebook_edit.desktop
-      - install -Dm644 -t /app/share/applications com.calibre_ebook.calibre.ebook_viewer.desktop
-      - install -Dm644 -t /app/share/applications com.calibre_ebook.calibre.lrfviewer.desktop
+
+      # Ensure that Calibre install script and the `xdg-utils` it uses will not
+      # report errors because they are unable to find any suitable destionation
+      # directory to place their files into
+      - mkdir -p "${FLATPAK_DEST}/bin" "${FLATPAK_DEST}"/share/{applications,desktop-directories,icons/hicolor,mime/packages}
+
+      # Run Calibre system integration scripts
+      - /app/lib/calibre/calibre_postinstall --root="${FLATPAK_DEST}"
+
+      # Rename installed Calibre .desktop files, app icons and appdata XMLs so that Flatpak will not
+      # refuse to use them
+      - python3 ./fixup-exports.py
```

Comment:
Can't this be done with regular calls to `desktop-file-edit` and a bunch of `mv` or `install` commands? The comment here suggests it should be. Hiding the process behind a Python script adds an overhead that I am not sure is necessary.

Reply:
It also edits the AppStream XMLs and adds aliases for the non-qualified Calibre AppStream names (that is `calibre-gui.appdata.xml` et al) and makes sure that the icon/desktop-file/AppStream names stay in sync. While likely not impossible, I don't think this would become much easier to read if using a shell script… (Note, that the actually interesting list of things is in lines 107 – 115 and 142 – 154, everything else is mostly boring support code.)
Additionally, the workaround in lines 107 – 115 will probably not be possible to replicate without the usage of XSLT from a shell script.

(As I encountered similar issues with other Flatpaks, I was also thinking about splitting this stuff into reusable utility with some Calibre-specific ruleset, but so-far all my other targets have accepted patches to generate the metadata in the places Flatpak wants them to instead, so the motivation for actually doing this was pretty low.)

For the time being hower, unless you feel really strongly about this, please just accept it mostly as-is and we can discuss how to improve things later. It's still a step-up from manually managing the entire metadata as it is.

Reply:
> unless you feel really strongly about this

It's great work, I'll deal with improvements later.

Just a few questions after a local build :
- Should the injected `<provides><id>calibre-gui.desktop</id></provides>` be forwarded upstream actually?
- There are a few mimetype files which are not exported:

```
Not exporting share/icons/hicolor/16x16/mimetypes/application-x-kindle-application.png, non-allowed export filename
Not exporting share/icons/hicolor/16x16/mimetypes/application-x-mobi8-ebook.png, non-allowed export filename
Not exporting share/icons/hicolor/16x16/mimetypes/application-lrf.png, non-allowed export filename
Not exporting share/icons/hicolor/16x16/mimetypes/text-lrs.png, non-allowed export filename
Not exporting share/icons/hicolor/16x16/mimetypes/application-x-topaz-ebook.png, non-allowed export filename
Not exporting share/icons/hicolor/16x16/mimetypes/application-x-mobipocket-ebook.png, non-allowed export filename
```

should we care?

Reply:
> It's great work, I'll deal with improvements later.

Please @ me about this then, I think we should have a separate discussion in the issues about this.

> Should the injected <provides><id>calibre-gui.desktop</id></provides> be forwarded upstream actually?

No, they are there to ensure that software stores consider our Flatpak'd Calibre and the upstream one the same application although we renamed it to the fuly-qualified name. Since the upstream packaging already refers to itself as `calibre-gui.desktop`, it wouldn't make sense for it to “provide itself” in this way.

> There are a few mimetype files which are not exported […] should we care?

It's only the icons of the media types registered by Calibre. Since the icon names need to start with `${APP_ID}.` and the names of media types Calibre registered do not, it means that these media types will be displayed using the icons shipped by Calibre and will use whatever generic file icon is provided by the current user theme instead.
Let's open an issue about this later, it's a very minor thing really and it'll need some research on how to fix it (*if* it is even fixable currently).

Reply:
… some research later:

We can fix the icons, by doing the following:

  1. Inject an `<icon name="${APP_ID}.${media_name}" />` tag into each registered media type ([reference](https://specifications.freedesktop.org/shared-mime-info-spec/shared-mime-info-spec-0.21.html#idm45624794561616)), so that the referenced icon starts with the AppID as required by Flatpak
  2. Rename each of the to-be-exported icons by prefixing them with `${APP_ID}.`

Again, I'd nominate doing this in a subsequent MR.

Reply:
Indeed, let's file a separate issue for these :+1:

Reply:
> I think we should have a separate discussion in the issues about this.

Thanks to your PR, I was able to assess the delta between what is produced after using `calibre_postinstall` (which I am convinced now yields a positive benefit) and what is necessary for a successful Flathub export.

I still believe the same changes could be achieved with a bunch of calls to `mv`, `desktop-file-edit` and `patch` (for the appstream metadata) which would avoid putting post-install logic wrapped in "mostly boring support code", as you said yourself. But since it's yet to be tested, and I can also be plain wrong about it, I will happily accept this PR :+1: 

Once again thanks @ntninja for your great work and @Erick555 for additional counsel.